pxr/imaging/plugin/hdRpr/volume.cpp

```diff
@@ -15,162 +18,107 @@ TF_DEFINE_PRIVATE_TOKENS(
-	};
-}
-
-
-GfVec3f colorLerp(const GfVec3f & color0, const GfVec3f & color1, const float & w)
-{
-	return color0 + (color1 - color0) * w;
-}
-
-
-GfVec3f temteratureToColor(const float & temperature)
+std::string findOpenVdbFilePath(HdSceneDelegate* sceneDelegate, const SdfPath & id)
 {
-	if (temperature <= tempColors.begin()->first)
-	{
-		return tempColors.begin()->second;
-	}
-	if (temperature >= tempColors.rbegin()->first)
+	for (auto it : sceneDelegate->GetVolumeFieldDescriptors(id))
 	{
-		tempColors.rbegin()->second;
-	}
-
-	auto tIt = tempColors.begin();
-	float tLess = tIt->first;
-	GfVec3f colorLess = tIt->second;
+		VtValue param = sceneDelegate->Get(it.fieldId, HdRprVolumeTokens->filePath);
 
-	for (++tIt; tIt != tempColors.end(); tIt++)
-	{
-		if ((int)temperature < tIt->first)
+		if (param.IsHolding<SdfAssetPath>())
 		{
-			float tMore = tIt->first;
-			GfVec3f colorMore = tIt->second;
-
-			float waight = (temperature - tLess) / (tMore - tLess);
-
-			return colorLerp(colorLess, colorMore, waight);
+			return param.Get<SdfAssetPath>().GetAssetPath();
 		}
-
-		tLess = tIt->first;
-		colorLess = tIt->second;
 	}
 
-	return GfVec3f();
-}*/
-
-size_t coordToIndex(const openvdb::Coord & coord, const openvdb::CoordBBox & activeBB, openvdb::Coord gridSize)
-{
-	const openvdb::Coord & v0 = activeBB.min();
-	const openvdb::Coord & v1 = activeBB.max();
-
-	float dx = ((float)(coord.x() - v0.x())) / (v1.x() - v0.x());
-	float dy = ((float)(coord.y() - v0.y())) / (v1.y() - v0.y());
-	float dz = ((float)(coord.z() - v0.z())) / (v1.z() - v0.z());
-
-
-	size_t xn = static_cast<size_t>(roundf(dx * gridSize.x()));
-	size_t yn = static_cast<size_t>(roundf(dy * gridSize.y()));
-	size_t zn = static_cast<size_t>(roundf(dz * gridSize.z()));
-
-	return (zn)*(gridSize.x()* gridSize.y()) + (yn)*(gridSize.x())+xn;
-};
-
+	return "";
+}
 
-std::string findOpenVdbFilePath(HdSceneDelegate* sceneDelegate, const SdfPath & id)
+void findNeededChannels(HdSceneDelegate* sceneDelegate, const SdfPath & id, bool &outNeedColor, bool &outNeedDensity, bool &outNeedEmissive)
 {
+	outNeedColor = false;
+	outNeedDensity = false;
+	outNeedEmissive = false;
 	for (auto it : sceneDelegate->GetVolumeFieldDescriptors(id))
 	{
 		VtValue param = sceneDelegate->Get(it.fieldId, HdRprVolumeTokens->filePath);
 
 		if (param.IsHolding<SdfAssetPath>())
 		{
-			return param.Get<SdfAssetPath>().GetAssetPath();
+			if (it.fieldName.GetString() == "color")
+				outNeedColor = true;
+			if (it.fieldName.GetString() == "density")
+				outNeedDensity = true;
+			if (it.fieldName.GetString() == "emissive")
+				outNeedEmissive = true;
 		}
 	}
 }
 
-template <class TPtr, class TValueOnIt>
-openvdb::CoordBBox computeGrigOnBB(TPtr grid)
+void findTemperatureOffsetAndScale(HdSceneDelegate* sceneDelegate, const SdfPath & id, float &outOffset, float &outScale)
 {
-	if (grid->empty())
+	for (auto it : sceneDelegate->GetVolumeFieldDescriptors(id))
```

Comment:
Could you please query VolumeFieldDescriptors once at Sync and pass them to all those "find*something" functions? This will allow us to avoid the unnecessary copy of the same vector every time you try to "find" something

Reply:
Wiil do.